### PR TITLE
Enhanced the voucher capabilities.

### LINKED ISSUE #13: Consider moving amount of Blanks to be minted to the voucher struct
Currently, the max amount of Blanks per member is set in the contract at 5, and a voucher can be redeemed for any amount up to 5 (and will error on an attempt to redeem it again).  If we want to issue vouchers for variable amounts of Blanks, we may want to move the amount to the voucher itself.

Issues this may cause:
1) will have to check for redeeming a voucher multiple times if we remove memberMaxMintcount
2) will have to deal with cases where someone waits to redeem a voucher when all Blanks are gone (this is the case in the current implementation as well, so may not be an issue, but raising it just in case)

### PR DESCRIPTION
**Pull request by asebesta**
- Added voucher expiration and token count.
- Added support to pause minting
 - Added support to update the maxMintCount

Resolves #13

### CODE REVIEW

File: contracts/BlankArt.sol [hunk 1]
```diff
@@ -114,7 +123,7 @@ contract BlankArt is ERC721, EIP712, ERC721URIStorage {
         else
             _base = _baseURIs[_baseURIs.length -1];
 
-        return string(abi.encodePacked(_base, Strings.toString(tokenId)));
+        return string(abi.encodePacked(_base, Strings.toString(tokenId), '.json'));
```
Comment (asebesta) [inline contracts/BlankArt.sol:126]: Added the .json extension, as it seems to require it for any tokens above 1, but let me know if there are other approaches here.

File: contracts/BlankArt.sol [hunk 2]
```diff
@@ -176,12 +196,17 @@ contract BlankArt is ERC721, EIP712, ERC721URIStorage {
     function redeemVoucher(uint256 amount, BlankNFTVoucher calldata voucher)
         public
         payable
-        returns (uint256[5] memory)
+        returns (uint256[] memory)
```
Comment (asebesta) [inline contracts/BlankArt.sol:199]: Made the return array dynamic since we can support more than 5 now.

File: lib/LazyMinter.js [hunk 3]
```diff
@@ -33,17 +33,21 @@ class LazyMinter {
...
    * @param {address} redeemerAddress the address authorized to redeem the voucher
+   * @param {ethers.BigNumber | number} expiration expiration date for the voucher, expressed in seconds since the Unix epoch
    * @param {ethers.BigNumber | number} minPrice the minimum price (in wei) that the creator will accept to redeem this NFT. defaults to zero
+   * @param {ethers.BigNumber | number} tokenCount the max number of tokens a voucher may redeem. defaults to five (5)
    *
    * @returns {BlankNFTVoucher}
    */
-  async createVoucher(redeemerAddress, minPrice = 0) {
-    const voucher = { redeemerAddress, minPrice }
+  async createVoucher(redeemerAddress, expiration, minPrice = 0, tokenCount = 5) {
```
Comment (asebesta) [inline lib/LazyMinter.js:42]: Currently defaulting to 5, but let me know if we want to remove that.

File: contracts/BlankArt.sol [hunk 4]
```diff
@@ -34,6 +34,8 @@ contract BlankArt is ERC721, EIP712, ERC721URIStorage {
     uint256 public maxTokenSupply;
     // cost to mint during the public sale
     uint256 public mintPrice;
+    // Enables/Disables voucher redemption
+    bool public active;
```
Comment (sjwhitmore) [inline contracts/BlankArt.sol:38]: nit: should we call this "voucherRedemptionActive" or something slightly more descriptive?

File: contracts/BlankArt.sol [hunk 5]
```diff
@@ -199,24 +224,34 @@ contract BlankArt is ERC721, EIP712, ERC721URIStorage {
         // make sure that the redeemer is paying enough to cover the buyer's cost
         require(msg.value >= (voucher.minPrice * amount), "Insufficient funds to redeem");
 
+        require(
+            amount <= voucher.tokenCount,
+            "Amount is more than the voucher allows"
+        );
+
+        // make sure voucher has not already been claimed. If true, it HAS been claimed
+        require(!voucherClaimed[_hash(voucher)], "Voucher has already been claimed");
```
Comment (sjwhitmore) [inline contracts/BlankArt.sol:233]: i just want to flag this for someone more knowledgable to doublecheck that this is the correct hash function to use? @fordN ?

**Review by sjwhitmore — COMMENTED** [review]
looks good to me!! going to leave to @fordN for a final pass.

File: contracts/BlankArt.sol (hunk 4 above)
Comment (asebesta) [reply contracts/BlankArt.sol:38]: Great feedback! I ended up using it for both redemption and minting, so at a minimum need to update the comment. If we want to restrict it to only redemption we can make the name more specific too!

File: contracts/BlankArt.sol (hunk 5 above)
Comment (asebesta) [reply contracts/BlankArt.sol:233]: Thanks - agreed! I'd definitely appreciate any feedback from someone with more experience there.

**Pull request merged**